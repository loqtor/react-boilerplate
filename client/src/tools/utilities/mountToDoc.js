/*
 * Utility to provide aXe with a fake DOM to allow for accessibility testing
 * on mounted components within our Jest tests
*/
import { mountWithIntl } from './enzymeIntlTestHelper';

let wrapper;

export default function mountToDoc(component) {
  const container = mountWithIntl(component);

  if (!wrapper) {
    wrapper = document.createElement('main');
    document.body.appendChild(wrapper);
  }

  wrapper.innerHTML = '';
  wrapper.appendChild(container.getDOMNode());

  return container;
}
