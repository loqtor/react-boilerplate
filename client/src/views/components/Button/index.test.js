import React from 'react';
import { shallow } from 'enzyme';
import axe from 'axe-core';
import toJson from 'enzyme-to-json';

import mountToDoc from '../../../tools/utilities/mountToDoc';

import { Button } from './index';

const setup = (render, props) => {
  const onClickSpy = jest.fn();

  const defaultProps = {
    disabledClass: '',
    isLinkButton: false,
    label: 'Continue',
    onClick: onClickSpy,
  };

  const component = render(<Button {...defaultProps} {...props} />);

  return {
    actual: component,
    onClickSpy,
  };
};

describe('<Button />', () => {
  it('renders correctly', () => {
    const { actual } = setup(shallow, {});
    expect(toJson(actual)).toMatchSnapshot();
  });

  it('renders correctly when disabled', () => {
    const { actual } = setup(shallow, { isDisabled: true });
    expect(toJson(actual)).toMatchSnapshot();
  });

  it('renders <a> tag correctly when is link', () => {
    const { actual } = setup(shallow, { isLinkButton: true, isRoute: false, url: '/' });
    expect(toJson(actual)).toMatchSnapshot();
  });

  it('renders <Link /> tag correctly when is link with route', () => {
    const { actual } = setup(shallow, { isLinkButton: true, isRoute: true, url: '/' });
    expect(toJson(actual)).toMatchSnapshot();
  });

  it('renders <Loader /> correctly when isLoading', () => {
    const { actual } = setup(shallow, { isLoading: true });
    expect(toJson(actual)).toMatchSnapshot();
  });

  it('has no accessibility violations', (done) => {
    const { actual } = setup(mountToDoc, {});
    const componentNode = actual.getDOMNode();

    axe.run(componentNode, (err, { violations }) => {
      expect(err).toBe(null);
      expect(violations).toHaveLength(0);
      done();
    });
  });
});
