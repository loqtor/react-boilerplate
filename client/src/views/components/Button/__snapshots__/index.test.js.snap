// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> renders <Link /> tag correctly when is link with route 1`] = `
<Link
  className=""
  id={null}
  onClick={[Function]}
  replace={false}
  role={null}
  title=""
  to="/"
>
  <span
    className=""
  >
    Continue
  </span>
</Link>
`;

exports[`<Button /> renders <Loader /> correctly when isLoading 1`] = `
<button
  className="is-loading"
  disabled={false}
  id={null}
  onClick={[Function]}
  role={null}
  title=""
  type="Submit"
>
  <div>
    <Loader
      className=""
      label=""
      type="circle"
    />
    <span
      className=" u-hiddenVisually"
    >
      Continue
    </span>
  </div>
</button>
`;

exports[`<Button /> renders <a> tag correctly when is link 1`] = `
<a
  className=""
  href="/"
  id={null}
  onClick={[Function]}
  rel="noopener noreferrer"
  role={null}
  target=""
  title=""
>
  <span
    className=""
  >
    Continue
  </span>
</a>
`;

exports[`<Button /> renders correctly 1`] = `
<button
  className=""
  disabled={false}
  id={null}
  onClick={[Function]}
  role={null}
  title=""
  type="Submit"
>
  <span
    className=""
  >
    Continue
  </span>
</button>
`;

exports[`<Button /> renders correctly when disabled 1`] = `
<button
  className=""
  disabled={true}
  id={null}
  onClick={[Function]}
  role={null}
  title=""
  type="Submit"
>
  <span
    className=""
  >
    Continue
  </span>
</button>
`;
